File: dynabuffers-java/src/main/kotlin/dynabuffers/ast/datatype/StringType.kt
package dynabuffers.ast.datatype

import dynabuffers.api.IRegistry
import dynabuffers.api.ISerializable
import dynabuffers.api.IType
import java.nio.ByteBuffer
import java.nio.charset.Charset

class StringType(private val options: StringTypeOptions) : IType, ISerializable {

    override fun size(value: Any?, registry: IRegistry) = 4 + str(value).toByteArray(options.charset).size

    override fun serialize(value: Any?, buffer: ByteBuffer, registry: IRegistry) {
        val encoded = str(value).toByteArray(options.charset)
        buffer.putInt(encoded.size)
        buffer.put(encoded)
    }

    override fun deserialize(buffer: ByteBuffer, registry: IRegistry): Any {
        val length = buffer.int
        val array = ByteArray(length)
        buffer.get(array)

        return String(array)
    }

    private fun str(obj: Any?) = obj.toString()

    data class StringTypeOptions(val charset: Charset)

}

File: dynabuffers-python/dynabuffers/ast/datatype/StringType.py
from dynabuffers.api.ISerializable import ISerializable


class StringTypeOptions:
    def __init__(self, charset: str):
        self.charset = charset


class StringType(ISerializable):
    def __init__(self, options: StringTypeOptions):
        self.options = options

    def size(self, value, registry):
        return 4 + len(str(value).encode(self.options.charset))

    def serialize(self, value, buffer, registry):
        encoded = value.encode(self.options.charset)
        buffer.putInt(len(encoded))
        buffer.put(encoded)

    def deserialize(self, buffer, registry):
        length = buffer.getInt()
        array = buffer.get(length)

        return array.decode(self.options.charset)
